src/Database/Expression/SelectExpression.php

```diff
@@ -0,0 +1,148 @@
+        $template = (!empty($modifiers) ? implode(' ', $this->getModifiers()) . ' ' : '') .
+            ($len === 1 ? '%s' : $prefix . '%s' . $suffix);
+        $parts = [];
+        foreach ($this->_conditions as $key => $part) {
+            if ($part instanceof Query) {
+                $part = '(' . $part->sql($generator) . ')';
+            } elseif ($part instanceof ExpressionInterface) {
+                $part = $part->sql($generator);
+            } else {
+                $p = $generator->placeholder('se');
+                $generator->bind($p, $part, $typeMap->type($key));
+                $part = $p;
+            }
+            if (strlen($part)) {
+                $parts[] = $part;
+            }
+        }
+
+        return sprintf(
+            $template,
+            implode(($conjPrefix ? ' ' : '') . $conjunction . ($conjSuffix ? ' ' : ''), $parts)
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function add($conditions, array $types = [])
+    {
+        $this->getTypeMap()->setTypes(array_merge($this->getTypeMap()->getTypes(), $types));
+
+        if (!is_array($conditions)) {
+            $this->_conditions[] = $conditions;
+        } else {
+            foreach ($conditions as $key => $condition) {
+                if (is_array($condition)) {
+                    $this->_conditions[] = new self($condition, [$this->getTypeMap()->type($key)]);
+                } else {
+                    if (!is_numeric($key) && ($condition == 'literal' || $condition == 'identifier')) {
+                        $this->_conditions[] = new IdentifierExpression($key);
+                    } else {
+                        $this->_conditions[] = $condition;
+                    }
+                }
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Adds a modifier to the set of modifiers to be applied to the expression.
+     *
+     * @param string $modifier The modifier to apply to the expression.
+     * @return $this
+     */
+    public function addModifier(string $modifier)
+    {
+        $this->modifiers[strtoupper($modifier)] = true;
+
+        return $this;
+    }
+
+    /**
+     * Removes the passed modifier if it exists within the set of modifiers.
+     *
+     * @param string $modifier The modifier to apply to the expression.
+     * @return $this
+     */
+    public function removeModifier(string $modifier)
+    {
+        $modifier = strtoupper($modifier);
+        if (isset($this->modifiers[$modifier])) {
+            unset($this->modifiers[$modifier]);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Method that returns the modifiers as either an associative array where
+     * the modifiers are keys or as an array where the modifiers are array
+     * values.
+     *
+     * @param bool $asValue Boolean determining return format.
+     * @return string[]
+     */
+    public function getModifiers(bool $asValue = true): array
```

Comment:
Do we need this parameter? It makes the return value inconsistent and doesn't appear to have any practical usage that the default value doesn't cover.